packages/router/__tests__/router-test.ts

```diff
@@ -1788,14 +1788,127 @@ describe("a router", () => {
       });
     });
+
+    it("navigates correctly using POP navigations", async () => {
+      let t = initializeTmTest();
+
+      let A = await t.navigate("/foo");
+      await A.loaders.foo.resolve("FOO");
+      expect(t.router.state.location.pathname).toEqual("/foo");
+
+      let B = await t.navigate("/bar");
+      await B.loaders.bar.resolve("BAR");
+      expect(t.router.state.location.pathname).toEqual("/bar");
+
+      let C = await t.navigate(-1);
+      await C.loaders.foo.resolve("FOO*");
+      expect(t.router.state.location.pathname).toEqual("/foo");
+
+      let D = await t.navigate("/baz", { replace: true });
+      await D.loaders.baz.resolve("BAZ");
+      expect(t.router.state.location.pathname).toEqual("/baz");
+
+      // POP to /
+      let E = await t.navigate(-1);
+      await E.loaders.index.resolve("INDEX*");
+      expect(t.router.state.location.pathname).toEqual("/");
+    });
+
+    it("navigates correctly using POP navigations across actions", async () => {
+      let t = initializeTmTest();
+
+      // Navigate to /foo
+      let A = await t.navigate("/foo");
+      await A.loaders.foo.resolve("FOO");
+      expect(t.router.state.location.pathname).toEqual("/foo");
+
+      // Navigate to /bar
+      let B = await t.navigate("/bar");
+      await B.loaders.bar.resolve("BAR");
+      expect(t.router.state.location.pathname).toEqual("/bar");
+
+      // Post to /bar (should replace)
+      let C = await t.navigate("/bar", {
+        formMethod: "post",
+        formData: createFormData({ key: "value" }),
+      });
+      await C.actions.bar.resolve("BAR ACTION");
+      await C.loaders.root.resolve("ROOT");
+      await C.loaders.bar.resolve("BAR");
+      expect(t.router.state.location.pathname).toEqual("/bar");
+
+      // POP to /foo
+      let D = await t.navigate(-1);
+      await D.loaders.foo.resolve("FOO");
+      expect(t.router.state.location.pathname).toEqual("/foo");
+    });
+
+    it("navigates correctly using POP navigations across action redirects", async () => {
+      let t = initializeTmTest();
+
+      // Navigate to /foo
+      let A = await t.navigate("/foo");
+      await A.loaders.foo.resolve("FOO");
+      expect(t.router.state.location.pathname).toEqual("/foo");
+
+      // Navigate to /bar
+      let B = await t.navigate("/bar");
+      await B.loaders.bar.resolve("BAR");
+      expect(t.router.state.location.pathname).toEqual("/bar");
+
+      // Post to /bar, redirect to /baz
+      let C = await t.navigate("/bar", {
+        formMethod: "post",
+        formData: createFormData({ key: "value" }),
+      });
+      let D = await C.actions.bar.redirect("/baz");
+      await D.loaders.root.resolve("ROOT");
+      await D.loaders.baz.resolve("BAZ");
+      expect(t.router.state.location.pathname).toEqual("/baz");
+
+      // POP to /bar
+      let E = await t.navigate(-1);
+      await E.loaders.bar.resolve("BAR");
+      expect(t.router.state.location.pathname).toEqual("/bar");
```

Comment:
This test would pass even with the wrong replace behavior because we have two /bars in a row right? The bug is that we'd lose the submitting entry.

```
// old broken behavior
GET /foo            -> [/foo]
GET /bar            -> [/foo, /bar (1)]
POST /bar, 30x /baz -> [/foo, /baz]
POP -1              -> [/foo]

// what we want is
GET /foo            -> [/foo]
GET /bar            -> [/foo, /bar (1)]
POST /bar, 30x /baz -> [/foo, /bar(1), /baz]
POP -1              -> [/foo, /bar]
```

Note we want `/bar(1)` not `/bar(2)`

With this test we don't know which one we got.

Reply:
Good catch 👍   Added assertions for that in [1495d99](https://github.com/remix-run/react-router/pull/8979/commits/1495d996dcab7d7a101c69007ac579b023391497)